src/main/java/org/owasp/esapi/reference/DefaultValidator.java

```diff
@@ -1224,11 +1226,17 @@ public boolean isValidURI(String context, String input, boolean allowNull) {
 				//and if the URI has any queries that also happen to match HTML entities, like &para;
 				//it will cease conforming to the regex we now specify for a URL.
 				isValid = p.matcher(canonicalizedURI).matches();
+			}else{
+				if(allowNull && inputIsNullOrEmpty ){
+					isValid = true;
+				}
 			}
 			
 		}catch (IntrusionException e){
 			logger.error(Logger.SECURITY_FAILURE, e.getMessage());
 			isValid = false;
+		} catch (URISyntaxException e) {
+			logger.error(Logger.EVENT_FAILURE, e.getMessage());
 		}
```

Comment:
In response to my comment #4 about considering other catches for Exceptions,
specifically about the regex pattern in Validator.URL, you wrote:

> I think the failure for a bad regex to happen should happen back in the beginning when we first bring in those regexs in the first place. I just toyed with making the URL pattern "(((" and I don't see any errors or exceptions... and I'm also pretty sure that I've seen an open issue that we allow people to enter invalid regexes.

I just checked a very simple case, and if you try something like

    Pattern.compile( "(" );

a PatternSyntaxException is thrown at runtime with the message

    Unclosed group new index <line#>

That said, I think the reason you don't get an exception throwun is where the Validation.URL gets compiled is via the DefaultSecurityConfiguration.getValidationPattern() and that method catches PatternSyntaxException, calls logSpecial(), and then just returns null (rather than the a Pattern object). (Of course it only does that if it doesn't find it already find the regex in the "pattern cache".)

What I think that means is if

    Pattern p = ESAPI.securityConfiguration().getValidationPattern( "URL" );

returns a null, we should interpret that as an invalid regex.

> The only way I could validate that the regex isn't bad here would be to manually compile a regex... and for multiple reasons I don't want to do that. Should we open a new feature request to validate regex syntax on bootup?

I think that's already being done via the call

    ESAPI.securityConfiguration().getValidationPattern( "URL" );

so either explicitly checking for 'p' equal to null right after that or explicitly catching a NullPointerException should allow us to detect an invalid regex for Validator.URL. If we do that, at least we would have an opportunity to log some regular message to wherever the normal ESAPI logging goes. (Logging done in DefaultSecurityConfiguration via logSpecial() simply to to System.out.)

-kevin